cmd/agent/daemon/state/events_pipeline.go

```diff
@@ -19,125 +20,278 @@ func (c *Controller) runEventsPipeline(ctx context.Context) error {
 	c.log.Info("running events pipeline")
 	defer c.log.Info("events pipeline done")
 
+	var currentEventsCount int
+	lastFlushedAt := c.nowFunc()
+	groups := map[uint64]*castpb.ContainerEvents{}
+
+	ticker := time.NewTicker(c.cfg.EventsFlushInterval)
+	defer ticker.Stop()
+
+	send := func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		batch := &castpb.ContainerEventsBatch{}
+		for _, group := range groups {
+			if len(group.Items) > 0 {
+				batch.Items = append(batch.Items, group)
+			}
+		}
+		if len(batch.Items) == 0 {
+			return
+		}
+
+		c.log.Debugf("sending events batch, events=%d, groups=%d", currentEventsCount, len(batch.Items))
+		if err := c.exporters.ContainerEventsSender.Send(ctx, batch); err != nil {
+			c.log.Errorf("sending events batch: %s", err)
+		}
+
+		// Reset state after data is sent.
+		for _, group := range groups {
+			group.Items = group.Items[:0]
+		}
+		lastFlushedAt = c.nowFunc()
+		currentEventsCount = 0
+	}
+
+	handleEvent := func(e *ebpftypes.Event) {
+		group, found := groups[e.Context.CgroupID]
+		if !found {
+			group = c.newContainerEventsGroup(e)
+			groups[e.Context.CgroupID] = group
+		}
+		pbEvent := &castpb.ContainerEvent{}
+		c.fillProtoContainerEvent(pbEvent, e, nil)
+
+		if c.enrichmentService.Enqueue(&enrichment.EnrichedContainerEvent{
+			Event:     pbEvent,
+			EbpfEvent: e,
+		}) {
+			return
+		}
+
+		group.Items = append(group.Items, pbEvent)
```

Comment:
This doesn't seem right, since events are send to the enrichment channel. This will cause duplicate events on the backend.

Reply:
Inside handleEvent if even should be enriched it's not added to the batch yet. But I see there is no test for this case, will add it.